RFCs/FS-1063-support-letbang-andbang-for-applicative-functors.md

```diff
@@ -0,0 +1,769 @@
+
+```fsharp
+// If both reading from the database or the file go wrong, the computation
+// can collect up the errors into a list to helpfully present to the user,
+// rather than just immediately showing the first error and obscuring the
+// second error
+result {
+  let! users = readUsersFromDb()
+  and! birthdays = readUserBirthdaysFromFile(filename)
+  return updateBirthdays users birthdays
+}
+```
+
+[Capriotti & Kaposi's paper](https://paolocapriotti.com/assets/applicative.pdf) introduces an example of creating an command line argument parser, where a single applicative can both statically generate help text for the parser, and dynamically parse options given to an application. [eulerfx](https://github.com/fsharp/fslang-suggestions/issues/579#issuecomment-309764738) has imagined an F# interpretation of that:
+
+```fsharp
+// One computation expression gives both the behaviour of the parser
+// (in terms of how to parse each element of it, what their defaults should
+// be, etc.) and the information needed to generate its help text
+opt {
+  let! username = Opt("username", (Some ""), Some)
+  and! fullname = Opt("fullname", None, Some)
+  and! id = Opt("id", None, readInt)
+  return User(username, fullname, id)
+}
+```
+
+With all of these examples, we can nest the applicative computation expressions inside other computation expressions to build larger and larger descriptions that cleanly separate the pure computation from its context.
+
+# Detailed Design
+[design]: #detailed-design
+
+This RFC introduces a desugaring for applicative computation expressions, much like that which exists for monads and related constructs.
+
+Example desugaring:
+
+```fsharp
+ce {
+    let! x = foo
+    and! y = bar
+    and! z = baz
+    return x + y + z
+ }
+```
+
+becomes
+
+```fsharp
+ce.Apply(
+    ce.Apply(
+        ce.Apply(
+            ce.Return(
+                (fun x ->
+                    (fun y ->
+                        (fun z ->
+                            x + y + z
+                        )
+                    )
+                )),
+            foo),
+        bar),
+    baz)
+```
+
+## Rules on keywords in applicative computation expressions
+
+To be accepted as an applicative computation expression (CE), the CE must be of the form `let! ... and! ... return ...`:
+
+* There must be at least one `and!`s after the `let!`, but there is no hard upper-limit on the number of `and!`s
+* No `let!`s may appear after an `and!` in the same applicative CE
+* No normal `let`s in between the `let!` and `and!`s
+* No usage of `yield` in place of `return`
+* `do!`, `match!` and other CE keywords are also not valid in an applicative CE
+* `use!` and `anduse!` may replace `let!` and `and!`, respectively, to indicate a resource that needs to be managed by the CE builder
+
```

Comment:
Additionally, though you had mentioned that you didn't account for errors yet, I think this RFC is at a place where this can be done. I think we should have a table of error messages that will be added when combining things incorrectly, and for the negative examples below, include what is emitted. In short, for every invalid combination we provide a code sample for, we should have an appropriate error message that we say will be emitted.

Reply:
You've done some of that here which is good.